bluemira/power_cycle/tools.py

```diff
@@ -31,3 +35,92 @@ def create_axes(ax=None):
     if ax is None:
         _, ax = plt.subplots()
     return ax
+
+
+@nb.jit
+def _nudge_check(x_last, x_this):
+    is_close = np.isclose(x_last, x_this, rtol=EPS)
+    is_decreasing = x_last > x_this
+    return is_close or is_decreasing
+
+
+@nb.jit
+def validate_monotonic_increase(x, strict_flag):
+    """Validate that vector is (strictly) monotonically increasing."""
+    for x_last, x_this in zip(x[:-1], x[1:]):  # noqa: RUF007, B905
+        if strict_flag and not (x_this > x_last):
+            raise ValueError("Vector is not strictly monotonically increasing.")
+        if not strict_flag and not (x_this >= x_last):
+            raise ValueError("Vector is not monotonically increasing.")
+
+
+@nb.jit
+def unique_domain(x: np.ndarray, epsilon: float = 1e-10):
+    """
+    Ensure x has only unique values to make (Domain: x -> Image: y) a function.
+
+    To be a function domain, x must be strictly monotonically increasing. So x
+    must start at least as a monotonically increasing vector. The function then
+    ensures strict monotonicity by nudging forward repeated values by a small
+    epsilon.
+
+    Epsilon must be small enough so that consecutive values can be considered
+    equal within the context/scale in which the function is defined.
+
+    The nudge forward for each non-unique element in x is (N * epsilon), given
+    N times, after the first appearance, that value has appeared in x before.
+    """
+    x = nb.typed.List(x)
+    new_x = nb.typed.List()
+    validate_monotonic_increase(x, strict_flag=False)
+    new_x.append(float(x[0]))
+    if len(x) > 1:
+        for x_this in x[1:]:
+            nudge = 0
+            new_x_this = x_this
+            needs_nudge = _nudge_check(new_x[-1], x_this)
+            if needs_nudge:
+                while needs_nudge:
+                    nudge += epsilon
+                    new_x_this = x_this + nudge
+                    needs_nudge = _nudge_check(new_x[-1], new_x_this)
```

Comment:
This could easily be a near infinite loop. Maybe something like the below at least makes the function loops a known max.
```suggestion
           if _nudge_check(new_x[-1], new_x_this):
                for _ in range(10):
                    nudge += epsilon
                    new_x_this = x_this + nudge
                    if _nudge_check(new_x[-1], new_x_this):
                        break
                else:
                    raise ValueError("Failed to create unique domain")
```

Reply:
There was a `not` missing in your suggestion, but it is implemented.